docs/howto/enumerations.md

```diff
@@ -0,0 +1,100 @@
+
+In the following example, extracted from the [coco2014 Croissant definition](https://github.com/mlcommons/croissant/blob/main/datasets/coco2014/metadata.json), the field `supercategory` is declared to be
+an enumeration using the `"ml:isEnumeration": true` property. This means that
+there is a finite number of `supercategory` values, and that it is expected
+that many rows will have the same value.
+
+https://github.com/mlcommons/croissant/blob/6c8eca2925191577a7101dcabd0a880aca03cae3/datasets/coco2014/metadata.json#L259-L271
+
+The field is of `dataType` `sc:Text`, but also has the type `sc:name`. This
+means that `supercategory` is expected to be a text, and that this text is
+the name of the enumeration value   ` naming itself (the
+supercategory), makes sense to humans, and could therefore be used as it in
+UIs.
+
+Although being straightforward, this approach of defining an enumeration has several limitations:
+
+1. The only way to know the cardinality of the enumeration (the number of
+   distinct values it can take), is to read the full `RecordSet` in which
+   the enumeration is defined, or to have data format native support for
+   enumerations.
+2. It doesn't allow specifying an enumeration value, a name and a semantic URL
+   together. If the underlying format allows it (e.g. Parquet), one might
+   declare an enumeration value (0, 1, 2, etc.) and a name, but no semantic
+   URL.
+
+For the above reasons, it is recommended to use a `RecordSet` to define an
+enumeration unless:
+
+1. the data format used by to store the `RecordSet` natively supports the
+  concept of enumerations, which should make it fast to list all possible
+  enumerations.\
+  or
+2. the `RecordSet` in which the enumeration is defined is known to be small.
+  This is the case in the example above, as the parent `RecordSet` is being
+  used to define another enumeration.
+
+### `ml:RecordSet` enumeration
+
+In the following example, also extracted from the [coco2014 Croissant definition](https://github.com/mlcommons/croissant/blob/main/datasets/coco2014/metadata.json), the `categories` `ml:RecordSet` is declared as being an enumeration:
+
+https://github.com/mlcommons/croissant/blob/6c8eca2925191577a7101dcabd0a880aca03cae3/datasets/coco2014/metadata.json#L237-L271
+
+That enumeration is then referenced from another `ml:RecordSet`, with a field
+referencing the enumeration `categories` `ml:RecordSet`:
+
+https://github.com/mlcommons/croissant/blob/6c8eca2925191577a7101dcabd0a880aca03cae3/datasets/coco2014/metadata.json#L284-L291
+
+When declaring an enumeration using a `ml:RecordSet`, that `ml:RecordSet` is
+expected to have at least one key field (here named `id`), which may be of
+various `dataType`s, but will most often be of type `sc:Integer` or `sc:Text`.
+
+As also described in the [labels howto](/howto/labels), the `ml:RecordSet` may
+also have fields of the following type:
+
+- `sc:name`, which may be the same field as the key field, if the key is of
```

Comment:
Again, this is using an sc property as a type... We don't support that at the moment. How about starting a discussion to figure out if we want to support this feature?

Reply:
I agree, this might be confusing. What about we create another type?
Something like "ml:Name"?

Reply:
I guess we have the same problem with "sc:url" which we use as a type in this same PR.
Should we have a type name something like "ml:UrlDefinition"?

Reply:
url is less of a problem, because we can use the existing class sc:URL. 

I couldn't find a class in schema.org that represent the fact that an object is a name. It also doesn't make much sense to do that...

If we want to go down the path of using existing properties from vocabularies, then I think we should specify a dataType (class) at the level of the RecordSet, and then the fields of the RecordSet can correspond to the properties of that class. We are already doing something along these lines for subFields: For instance, when we say that we have a field of type sc:GeoCoordinates, the subFields are latitude and longitude, which correspond to properties of that class.

Having DataType at the level of the RecordSet would also make it easier to define simple (single-field) enums / RecordSets by just specifying a simple type and not declaring fields.